src/codeCache.cpp

```diff
@@ -311,3 +307,60 @@ size_t CodeCache::usedMemory() {
     }
     return bytes;
 }
+
+bool CodeCache::isValidHandle(void* handle) const {
+#ifdef __linux__
+    struct link_map* map;
+    // validate that the current loaded library is the same library that was observed during the /proc/self/maps processing
+    if (handle != NULL && dlinfo(handle, RTLD_DI_LINKMAP, &map) == 0) {
+        return _image_base == (const char*)map->l_addr;
+    }
+    return false;
+#else
+    return handle != NULL;
+#endif
+}
+
+UnloadProtection::~UnloadProtection() {
+    if (_lib_handle) {
+        dlclose(_lib_handle);
+    }
+}
+
+UnloadProtection::UnloadProtection(CodeCache *cc) {
+    _protected_cc = cc;
+    _lib_handle = nullptr;
+    _valid = false;
+
+    if (Symbols::isMainExecutable(cc->_image_base, cc->_max_address) || Symbols::isLoader(cc->_image_base)) {
```

Comment:
`OS::isMusl()` check is needed here, too. `dlopen` protection is not applicable on Alpine.

Reply:
The check is required, otherwise `dlopen()` fails on musl if the underlying fail has been deleted. `dlopen` protection is useless on musl anyway, since libraries cannot be unloaded.

Reply:
45e962273959e117891b4b7d4adb73ade8d063fd